derp/derp_server.go

```diff
@@ -959,6 +965,20 @@ func (c *sclient) setPreferred(v bool) {
 	}
 }
 
+// expMovingAverage returns the new moving average given the previous average,
+// a new value, and an alpha decay factor.
+// https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average
+func expMovingAverage(prev, new_value, alpha float64) float64 {
+	return alpha*new_value + (1-alpha)*prev
+}
+
+// recordQueueTime updates the average queue duration metric after a packet has been sent.
+func (c *sclient) recordQueueTime(enqueuedAt time.Time) {
+	elapsed := float64(time.Since(enqueuedAt).Milliseconds())
+	newAverage := expMovingAverage(c.s.averageQueueDuration.Value(), elapsed, 0.1)
```

Comment:
this atomic read + atomic write can lose updates when called concurrently between different clients.

you really want some sort of CAS loop here, which expvar doesn't give you. So you'd probably want to use an int64 with sync/atomic yourself, and then use `expvar.Func` to read the varz out like some of the others in Server.ExpVar